### PR TITLE
Prepare project to use Zanata for translations

### PR DESCRIPTION
**Pull request by KurzedMetal**
I realized that with the recent refactors I made, I was changing some translation strings and they needed to be re-translated. Suddenly I realized some translations haven't been updated in a while and the project don't have an easy way to receive translation contributions without fiddling with gettext tools... Let's be real, potential contributors to translatation are pushed off if they need to actually fiddle with tons of tools instead of actually translating, most of them aren't even tech savvy or devs...

I always thought that web tools for translation like Transifex were super cool and wanted to configurate one of these services for CDDA Launcher. I checked Transifex (which is actually used by CDDA itself too), but they require an "approved" registration for open source projects as the service is usually paid. So I googled around and I found Zanata, which seems a similar open source service with less registration hassle.

I already configured everything the project needs to push and get translation from this page.
The workflow would be:
* You code stuff in CDDA Launcher, create or update translatable strings.
* Run `setup.py zanata_expush` to extract strings and push them to Zanata service
* Translators register on Zanata (free account and painless short steps)
* Translators do their work on a nice web interface with no tool required.
* You register, you approve them their translations (or not)
* Run `setup.py zanata_pull --approved-only` to get all the strings you approved in the web interface. Or no `--approved-only` and you get everything anyways.
* At this point, if you run `setup.py create_installer` it would include the new translations (no need to do anything else).

I already setup a temporary page for the project, so you could see it, register and join to check ([cdda-launcher-test in Zanata](https://translate.zanata.org/iteration/view/cdda-launcher-test/main)). But if you want you can register your own and just change the configuration files (just so you can get the ownership of it, since it's your project)

If you wanna create your own, you only need to: create an account, create a project, add a version, [update the `zanata.xml` file and configure your client](http://docs.zanata.org/en/latest/client/configuration/) (take a good look at that link, it's how you setup the client in your PC) and run `setup.py zanata_expush`. If you want to boostrap your new project and upload existing translations too you need to execute `setup.py zanata_push --push-translations` too (this is only for an initial setup, because it *overwrites* translations in the server)

A note worth mentioning: after I had a good progress in the configuration and almost everything working, I noticed some bad news around, Zanata was an open source project sponsored by Red Hat, but it seems they pulled the main devs from the projects a few months ago. The Zanata project seems mature and still hosted, so at least for now it works and it's better than you had before. But if you wanna try other service or maybe try getting a Transifex account for this project, maybe I could help setting it up too.

### CODE REVIEW

**Comment by remyroy** [comment]
That seems like a nice idea. Let me check the current state of Zanata.

**Comment by remyroy** [comment]
I contacted Transifex and they said that we qualify for an Open Source license. We can use their services for free. What do you think? Should we go with them? They say they have the largest translation community.

**Comment by KurzedMetal** [comment]
Yes, Transifex is probably better. I doubt the Zanata project is going down any time soon, and the feature set does seem currently stable, but it would be better to be in the safer side.

I'll check some Transifex documentation and see how we can push/pull Translations (if they even have such an API, which I bet they do)

**Comment by KurzedMetal** [comment]
Can you create an organization / project for CDDA Launcher and invite me?
Seems that configuring project requires those to be already created ([`tx init`](https://docs.transifex.com/client/init) with their [client](https://docs.transifex.com/client/installing-the-client)).

Also, If you give me "Project Maintainer" Role could probably let me do some tests (I wanna add some new `setup.py` commands) and push initial data.

From what I read so far, the workflow is similar:
* [install the client](https://docs.transifex.com/client/installing-the-client)
* initial setup with: `tx init` (this creates a `~\.transifexrc` with you auth token too)
* push untranslated strings with `tx push --no-interactive -s` (add `-t` for translations too, probably useful only in the initial setup)
* pull translated strings with `tx pull --no-interactive -a --mode onlyreviewed` (or `onlytranslated` if you wanna accept every translation from contributors without reviews)

**Comment by remyroy** [comment]
I'll create the organization in the next few minutes.

**Comment by remyroy** [comment]
What is your Transfinex username?

**Comment by KurzedMetal** [comment]
Same as here:[ KurzedMetal](https://www.transifex.com/user/profile/KurzedMetal/)

**Comment by remyroy** [comment]
Transifex seems to have some kind of Github integration where Transifex can push and pull directly on Github. I'm trying to activate it. That might help. If not let me know.

**Comment by KurzedMetal** [comment]
Yes, you can configure Transifex to check changes in a hosted file, and they check it twice a day.
But I was trying to configure it so you can push the translations whenever you want on demand.

Your choice, both will probably work, you can probably even use both I guess.

**Comment by remyroy** [comment]
I was just trying to setup the GitHub integration but I don't think I succeeded. I see that you started adding some resources. I'll let you explore more of it and come back later.

**Comment by KurzedMetal** [comment]
Yes, I was testing the `tx` commands manually and they worked great. It pushed everything already except for 3 plural Russian strings because it seems Transifex expected more strings for them.

I'll use those `tx` commands to create `setup.py` commands, those will be handy, and you can use it whenever you want, I'm pretty sure they won't interfere from configuring automatic updates too, if that's what you want.

Anyways, I should have something soon, as it's very similar as what I had with Zanata.

**Comment by remyroy** [comment]
Good stuff. There are various ways of configuring the GitHub integration. I've set it to create pull requests instead of directly pushing into the repository.

We might also need to tweak the .gitignore file to keep that pot file around.

**Comment by KurzedMetal** [comment]
I created the new commands for Transifex, so now, the workflow should be like:
* You code stuff in CDDA Launcher, create or update translatable strings.
* Run `setup.py translation_expush` to extract strings and push them to Transifex service
* Translators navigate to the [project Dashboard](https://www.transifex.com/cdda-game-launcher/cdda-game-launcher-project/dashboard/) and click "Help translate CDDA Game Launcher" to register.
* Translators do their work on a nice web interface with no tool required.
* Someone review the translations (or not)
* Run `setup.py translation_pull --reviewed-only` to get all the reviewed strings in the web interface. Or no `--reviewed-only` and you get every translated anyways (reviewed or not).
* At this point, if you run `setup.py create_installer` it would include the new translations (no need to do anything else).

I also left around the code/configuration for Zanata, in case you want to use them in the future, but I could purge all reference of it if you want, your choice.

**Comment by KurzedMetal** [comment]
I found a minor issue:
After `translation_pull`ing all `.po` files with translations, I noticed that Transifex marks it the files as "Fuzzy" and Babel refuses to compile them into `.mo`. That leaves the Launcher without the compiled language files it needs and switching languages on UI has no effect.

Trying to figure out why Transifex are marking them as Fuzzy. The french `.po` still gets that flag even after I marked all entries as "Reviewed" and have like 96% translated strings.

**Comment by KurzedMetal** [comment]
Ok... I think I found some info about this...

Transifex is giving us a PO file with "fuzzy" flag on the header, which means it may need to be reviewed and completed (for example there is a `FIRST AUTHOR <EMAIL@ADDRESS>` line and a `"Report-Msgid-Bugs-To: EMAIL@ADDRESS\n"` field that could be tweaked). BTW, I checked already and it doesn't seem to be any way to change this flag or fields from Transifex.

But, it seems having "fuzzy" flag on the header makes Babel think the whole translation is fuzzy and skip all the file content while compiling it, even when the text translations themselves aren't fuzzy.

I found that `babel.compile_catalog` has a `--use-fuzzy` flag to avoid skipping fuzzy lines, which IMO isn't that bad because I'd totally prefer a fuzzy translated string than a not translated at all.
Also, translation correctness should be handled with reviews on Transifer from now on and Transifex doesn't seem to use "fuzzy" on individual messages.

**Comment by KurzedMetal** [comment]
BTW, I also finished translating the Launcher to Spanish on Transifex, which is my main language :)

I left ~50 words out because I'm not exactly sure how to translate a few words like "Mods", "Soundpacks" and "Tilesets", which don't have a direct translation, but I'll figure out something sooner or later, I just want to think about them a bit.

I also probably missed a accents (I'm terrible with them) but grammar and wording should be fine.

**Comment by remyroy** [comment]
That's really great for the Spanish translation! Have you tested it with the launcher to see if it shows properly in the different tabs?

Regarding the translation check, do you know which one it is? Here is the list I can see:

checks

**Comment by remyroy** [comment]
Regarding fuzzy translation, I think we should accept them and add that command line argument for babel to process them properly.

**Comment by KurzedMetal** [comment]
I'd set to Warning (at least for now), the "typographical characters" and "html tags" and the "setup placeholder delimiters" and add configure the last one to support for python string format params `{...}`.

**Comment by remyroy** [comment]
Good. The new settings should be in place now.

**Comment by remyroy** [comment]
Do you know how to assign a translator to a specific language on Transifex? I'm trying to assign myself French and I would assign you Spanish but I can't seem to find that action. All I see is no translator for each language.

**Comment by KurzedMetal** [comment]
Check if you can see anything in the Team management section, I think the link is https://www.transifex.com/cdda-game-launcher/teams/101126 I just don't have permissions (I get a 403 forbidden)

I got there by going to Language -> Spanish -> View members -> and clicking the "CDDA Game Launcher Translation Team" hyperlink... some images may be better:

Languages section and click any language...
image

Then click View members
image

You may be able to do something in that screen already, I only see empty lists. All I see is "Organization administrators and team managers can coordinate this team."
image

**Comment by remyroy** [comment]
I managed to add myself as the Coordinator for French and you as the Coordinator for Spanish. I'm not sure we need all these roles but it removes the `no translator` tag for those languages.

I would prefer to remove the Zanata stuff if we are to use Transifex. If you want to keep the Zanata stuff, just create a seperate branch in your GitHub repo and we'll be able to revisit it later on if we need to.

**Comment by remyroy** [comment]
Regarding the MIT license, I know there are translated version of it but I'm not sure from a legal standpoint and from a usability standpoint what's the best approach. In any case, I don't think it's wrong to leave it in English all the time.

**Comment by KurzedMetal** [comment]
I just googled it, there's no official translation of the Licenses, there are unofficial ones, mostly to give users awareness, I don't think it's necessary, I did add a new string saying "This software is distributed under the MIT License" maybe we can expand it with "This is free software, completely free of use and distribution under the terms of the MIT License" which should be enough information for normal users (who reads 3 paragraph of legal stuff :P)

**Comment by KurzedMetal** [comment]
BTW, all these small changes on strings were improvements I found while translating Spanish and I felt that could reduce hazzle and improve maintainability of the translations.

I just completed the Spanish translation to 100%, so I doubt I'll do any other modifications for now, and the commands for pushing / pulling worked correctly. I swear I won't add any other commit in this PR 😄 If I have anything else I'll create a new one.

You are good to do a final revision *without more changes* :P.

**Comment by remyroy** [comment]
Yeah, that is a good idea. We could only translate that single simple phrase for everyone and keep the canonical license in English.

**Comment by KurzedMetal** [comment]
Ok, I'm dead serious now :P that was the last push. ;P

**Comment by remyroy** [comment]
There was an option to enable anyone to join the team. I checked it.
settings

**Comment by KurzedMetal** [comment]
> I would prefer to remove the Zanata stuff if we are to use Transifex. If you want to keep the Zanata stuff, just create a seperate branch in your GitHub repo and we'll be able to revisit it later on if we need to.

I totally missed that before

**Comment by KurzedMetal** [comment]
It seems I hit a [babel bug](https://github.com/python-babel/babel/issues/35) when using `translation_pull` with the change where I added the MIT License summary. I used the text "100% free" and it seems babel is reading that as a string format "%f" (yes even if it has a space between `%` and `f` because it's actual valid python syntax).

I'm getting the error ```error: cddagl\locale\es\LC_MESSAGES\cddagl.po:470: incompatible format for placeholder 1: 'f' and 's' are not compatible```.

I'd really hate workaround it by changing the message, I was checking if I could pass some option to babel to avoid this but I couldn't find it so far.

**Comment by remyroy** [comment]
I'm confused about this part in `i18n.py`:

```python
def load_gettext_locale(locale_dir, locale, domain='cddagl'):
    """Load gettext into the application with specified locale file.

    Fallback to default untranslated strings if locale file is not found.
    """
    if locale == 'en':
        return load_gettext_no_locale()
```

Why are we not loading all the locales if en is selected?

Why is 'en' is value for when we cannot find the locales? Would it not be possible for 'en' to be a normal value and we still need to load the locales at that point in load_gettext_locale?

**Comment by KurzedMetal** [comment]
Until now, the code depended on an empty `po` to have english translations.

When 'en' was loaded, it was actually provided with an empty file (well, actually not "empty", but with no translations). That made the code use the translations found in the source code instead of the non-existent translations in the empty `.po`.

With two small changes, I made the code not depend on the empty `.po` file for english to actually load English translations, it uses `load_gettext_no_locale()` which is exactly the same as giving it an empty file (meaning it directly uses the strings in the source code).

Basically, that change add two lines of code to avoid having an useless empty file.

**Comment by remyroy** [comment]
English shows twice in the Language settings. It should only show once.

double-english

**Comment by KurzedMetal** [comment]
That issue happened because you still have the empty english `.mo` file around when I removed it locally. I already deleted the `.po` file from the repo, but there's nothing I could do for the `.mo` because it's git-ignored.

Since the `.mo` is ignored, I can't actually delete it or anything, so I improved the `get_available_locales` to avoid duplicates. But in reality, the issue is that you have a file that shouldn't be there, and that `git clean -X -f` would delete.

**Comment by remyroy** [comment]
Alright, good to know.

**Comment by remyroy** [comment]
Seems good. Merging.

**Comment by chriseich21** [comment]
You may want to setup a Patreon instead of asking for cryptos, considering their price lmao. But great work on the continued support and really glad that autoupdate is working now!

**Comment by chuanyueyouxia** [comment]
May I translate it into Chinese？I have already requested on transifex.

**Comment by remyroy** [comment]
@chuanyueyouxia Sure! Let me check it out on Transifex and setup Chinese there.

**Comment by remyroy** [comment]
@chuanyueyouxia There are [various forms of Chinese in the localization world](https://stackoverflow.com/questions/4892372/language-codes-for-simplified-chinese-and-traditional-chinese). I know that you requested *zh_CN* but would it make sense to do others as well like one of these:

* **zh_SG** for Singapore
* **zh_TW** for Taiwan
* **zh_HK** for Hong Kong
* **zh_HANS** for (generic) simplified Chinese characters
* **zh_HANT** for (generic) traditional Chinese characters

I know that there can be some subtle differences from region to region but it might not be worth to do them. For instance, I'm French Canadian. We have this fr_CA code here but there are multiple french variants like fr_FR (French from France), fr-BE (French in Belgium), etc. Since there are not enough differences, I decided to go for the generic fr code. Let me know what you think about the Chinese language for this.

Also, I would need the email address or the username you used to create your profile on Transifex so I can add you to the Chinese translation team as a translator.

Thanks

**Comment by remyroy** [comment]
I think I was able to add you to the correct team with the proper role on Transifex now so I will not need your username or email.

If you need help with the translation or if there is something you don't understand, let us know. It can be somewhat difficult with the various codes and markups that are used. We'll be happy to help you out @chuanyueyouxia !

**Comment by chuanyueyouxia** [comment]
@remyroy Thanks!
Oh, I see.There are many forms of Chinese in the localization world.Chinese can be written as simplified Chinese and traditional Chinese.
IN CDDA game, Chinese zh_CN is simplified Chinese, Chinese zh_TW for Taiwan is traditional Chinese.
Maybe we can use general code?

**Comment by remyroy** [comment]
I think **zh_HANS** might be the generic simplified one that we are looking for. I'll switch to that one on Transifex and add you there on that team too.

**Comment by remyroy** [comment]
@chuanyueyouxia The switch to **zh_HANS** from *zh_CN* is completed on Transifex. You should be good now to start translating.

Thanks

**Comment by chuanyueyouxia** [comment]
Thanks a lot, @remyroy . You've been great！
